### PR TITLE
feat(ast_tools): output typescript to a separate package

### PR DESCRIPTION
**Pull request by ottomated**
Part of #6347.

Moves typescript logic from derive_estree into a new ast_tools generator.

### CODE REVIEW

**Comment by ottomated** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6755?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6755** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6755?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6687** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6687?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6686** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6686?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6685** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6685?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6683** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6683?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ottomated and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by Boshen** [comment]
👍 nice work

**Comment by Boshen** [comment]
Our ts types came from typescript-eslint, I would model them as such https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/ast-spec/src/expression/ArrayExpression/spec.ts

but this is for another topic.

**Comment by overlookmotel** [comment]
I have taken the liberty of combining the changes from #6687 into this PR, since neither PR stands alone, and it was confusing trying to review each in isolation. I've preserved the separate commits from both PRs.

I've added a commit to remove unused dependencies, which unblocks the "autofix" CI job.

@ottomated You will need to force pull the branch to your local machine, and then run `gt track` to get Graphite back in sync.

This is great, but there are a few problems:

#### Autofix

The "autofix" CI job automatically reformats all `.d.ts` files in the repo with `dprint`. But the "AST changes" CI job checks that generated files have not been altered manually. So it's freaking out because it sees the changes "autofix" made.

Longer term we should run `dprint` on the `.d.ts` in `ast_tools`, so it's formatted in standard style. But that can come later. Simplest way to solve the problem for now is to add the generated `.d.ts` file to autofix's ignore list:

https://github.com/oxc-project/oxc/blob/d2b8f505f136bbad0b3abffa3a8664a9439b305a/.github/workflows/autofix.yml#L4-L8

#### WASM types

This PR generates the `.d.ts` file, but doesn't link it into `wasm-bindgen`. So the types are now gone from the WASM package. This breaks that package, so this PR can't be merged as it stands (CI always needs to be green).

I suggest in this PR:

* Remove the new `oxc-types` package.
* Instead output the combined types in an `.rs` file as a single `const TS_APPEND_CONTENT` statement.
* Import that module into `oxc_ast/src/serialize.rs`.

That'll get this PR into a state in which it can be merged. Leave figuring out how to share the same type defs between `oxc-types` and WASM package to the next PR. *At that point*, re-introduce the `oxc-types` package.

#### Missing custom types

The generated `.d.ts` file is missing type defs for `FormalParameterRest`, `JSXElementName` etc which are currently defined in `const TS_APPEND_CONTENT` statements. I know you're aware of this, but it needs to be resolved before we can switch over to the generated `.d.ts`.

Sorry if this all feels like bureaucracy! It's annoying sometimes, but each individual PR (even in a stack) should not leave any crates / packages in a broken state. The advantage of stacked PRs is you can move fast (no need to wait for previous PR to get merged before starting on the next one), but the trade-off is that you need to consider the order you do things in, to keep CI green on every PR.

**Comment by ottomated** [comment]
@overlookmotel thanks for the PR juggling, I'm well aware that this PR is not done and have marked it as draft for clarity.

**Comment by ottomated** [comment]
@overlookmotel everything should be addressed now!

File: tasks/ast_tools/src/generators/typescript.rs
Comment (overlookmotel) [inline tasks/ast_tools/src/generators/typescript.rs:33]: Nit: Weird indentation.

File: tasks/ast_tools/src/generators/typescript.rs
Comment (overlookmotel) [inline tasks/ast_tools/src/generators/typescript.rs:57]: Nit.

```suggestion
// Untagged enums: `type Expression = BooleanLiteral | NullLiteral;`
// Tagged enums: `type PropertyKind = 'init' | 'get' | 'set';`
```

File: tasks/ast_tools/src/generators/typescript.rs
Comment (overlookmotel) [inline tasks/ast_tools/src/generators/typescript.rs:170]: Delete this dead code. Dprint is probably what we want to use anyway, since that's the tool we use to format all other JS/TS code in the repo.

**Review by overlookmotel — COMMENTED** [review]
All looks good. Very nice! My comments below are just nits.

The one remaining question is the name of the types package. We don't control the `@oxc` npm scope unfortunately, and Oxc's NPM packages are scattered around a bit.

We could use either a scope-less package name `oxc-types` or `@oxc-project/types`.

@Boshen?

File: tasks/ast_tools/src/generators/typescript.rs
Comment (ottomated) [reply tasks/ast_tools/src/generators/typescript.rs:170]: Only problem with dprint is it doesn't turn `(foo) | null` into `foo | null`.

**Comment by overlookmotel** [comment]
@Boshen I've reviewed all the Rust code, but could you please take a look at the package publishing + github workflow stuff?

File: tasks/ast_tools/src/generators/typescript.rs
Comment (overlookmotel) [reply tasks/ast_tools/src/generators/typescript.rs:170]: I was going to ask. Can we just output `foo | null` to start with?

I imagine in some cases `foo` *needs* wrapping in brackets, but I don't know TS well enough to know in what circumstances. Could we use a quick-and-dirty approach of wrapping only if it contains `{` or something?

Comment (ottomated) [reply tasks/ast_tools/src/generators/typescript.rs:170]: Yeah I went through and it doesn't seem to cause any issues here. If we were using prettier it would just be a no-brainer to include it but since dprint can't handle it it makes sense to remove the parentheses.

Comment (ottomated) [reply tasks/ast_tools/src/generators/typescript.rs:170]: I accidentally committed this to #6796, still getting used to graphite, sorry.

File: tasks/ast_tools/src/generators/typescript.rs
Comment (overlookmotel) [reply tasks/ast_tools/src/generators/typescript.rs:33]: Ah I think it's just a tabs-vs-spaces thing. Probably some tabs have crept in, so it renders weirdly if you have your IDE set to render tabs as 4 spaces. Our convention is to use spaces instead of tabs.

File: tasks/ast_tools/src/generators/typescript.rs
Comment (overlookmotel) [reply tasks/ast_tools/src/generators/typescript.rs:170]: No problem. It's just a nit anyway!

**Comment by Boshen** [comment]
> @Boshen I've reviewed all the Rust code, but could you please take a look at the package publishing + github workflow stuff?
> 
> Also, see above my question about package name for the types. `oxc-types` or `@oxc-project/types`?

I'll address these after this PR is merged.

**Comment by Boshen** [comment]
Let me make a new release for `oxc` before merging this.

**Comment by ottomated** [comment]
Could also be @oxc-ast/types or something along those lines

File: crates/oxc_ast/src/ast/types.d.ts
Comment (Boshen) [inline crates/oxc_ast/src/ast/types.d.ts:1]: This file seems out of place.

Comment (ottomated) [reply crates/oxc_ast/src/ast/types.d.ts:1]: Yeah this holds the manual implementation of any types that require it. @overlookmotel suggested putting it in the oxc_ast crate but it could go in ast_tools as well.

Comment (overlookmotel) [reply crates/oxc_ast/src/ast/types.d.ts:1]: I do think it should go in `oxc_ast`, so it's co-located with the rest of the AST serialization logic, rather than hidden away in `ast_tools`.

Perhaps it'd be better if we moved `serialize.rs` into a separate folder, and put this `types.d.ts` in that folder too. But I suggest we tackle that in a follow-up PR.

The intent is that this file is only a temporary measure anyway. Once `#[estree]` gets a bit more fully-featured and can generate types for these "special cases" itself, this file will be removed again.

**Review by overlookmotel — APPROVED** [review]
Boshen said he'd tackle the package release stuff after this is merged. Oxc 0.33.0 has been released. So I think this is good to merge now. Ditto #6796 and #6836.

@Boshen I'll leave it to you to merge, in case I've missing something.

**Pull request merged**
